compile.c

```diff
@@ -11606,17 +11609,37 @@ static void
 ibf_dump_object_symbol(struct ibf_dump *dump, VALUE obj)
 {
     VALUE str = rb_sym2str(obj);
-    VALUE str_index = ibf_dump_object(dump, str);
 
-    ibf_dump_write_small_value(dump, str_index);
+    long encindex = (long)rb_enc_get_index(str);
+    long len = RSTRING_LEN(str);
+    const char *ptr = RSTRING_PTR(str);
+
+    if (encindex > RUBY_ENCINDEX_BUILTIN_MAX) {
+        rb_encoding *enc = rb_enc_from_index((int)encindex);
+        const char *enc_name = rb_enc_name(enc);
+        encindex = RUBY_ENCINDEX_BUILTIN_MAX + ibf_dump_object(dump, rb_str_new2(enc_name));
+    }
+
+    ibf_dump_write_small_value(dump, encindex);
+    ibf_dump_write_small_value(dump, len);
+    IBF_WP(ptr, char, len);
```

Comment:
It seems to me you shouldn't need to change the binary format. This part is basically `ibf_dump_object_string()`. You should be able to read get at the string buffer in the object list when loading.

Reply:
> It seems to me you shouldn't need to change the binary format.

Is it a big deal though? I mean the ISeq format isn't stable AFAIK, and totally not interchangeable between ruby versions etc. So might as well skip the string "header", no?

Reply:
ibf_dump_object deduplicates though. So your change might end up making the output bigger.

Reply:
Hum. I haven't noticed that. I'll have a look.

Reply:
Ok, I think I see what you mean now, and yes in theory what you suggest could be more compact.

However in practice Symbols rarely share their backing strings:

```ruby
>> :foo.name.equal?("foo".freeze)
=> false
```

This is because their encoding don't match. Symbols default to ASCII, and strings to UTF-8:

```
>> puts ObjectSpace.dump(:foo.name)
{"address":"0x7fa4780c91c0", "type":"STRING", "class":"0x7fa4748c2858", "frozen":true, "embedded":true, "fstring":true, "bytesize":3, "value":"foo", "encoding":"US-ASCII", "memsize":40, "flags":{"wb_protected":true, "old":true, "uncollectible":true, "marked":true}}
>> puts ObjectSpace.dump("foo".freeze)
{"address":"0x7fa47490ce08", "type":"STRING", "class":"0x7fa4748c2858", "frozen":true, "embedded":true, "fstring":true, "bytesize":3, "value":"foo", "encoding":"UTF-8", "memsize":40, "flags":{"wb_protected":true, "old":true, "uncollectible":true, "marked":true}}
```

See also https://bugs.ruby-lang.org/issues/15940

So in the end, I don't think we gain much. What do you think?

Reply:
My main objection is to the code duplication. It seems you can call `ibf_dump_object_string()` instead of duplicating it? Seems exactly the same.

Reply:
Done.